### PR TITLE
feat: update minimal supported version of Node.js to v20

### LINKED ISSUE #601: Upgrade to newer version of node.js
Currently project works under Node 18 which reached its EOL on April 30, 2025. It will no longer receive security updates or maintenance releases.

Plan it to upgrade to version 22 or 24 (current LTS).

Note: this won't be an easy task since there are brekable changes.

Refs: https://nodejs.org/en/blog/announcements/node-18-eol-support

### PR DESCRIPTION
**Pull request by bliuchak**
This PR upgrades minimal supported version of Node.js to v20.

## Why Node.js 20?

I think for us it makes sense to support last 3 LTS versions of node. As of today they are: 24, 22 and 20.

Also I'll keep tests at Github Actions for Node.js 16, 18 and 20 (will be added) to make sure we're not breaking too much things. At least right now. However this might change in next major release.

## Migration Flow

Migration from Node.js 18 to 20 wasn't complicated. Here are the main issues which we've faced:
- Node.js 20 enables `keep-alive` by default.  This should be turned off in tests explicitly.
- Some tests should be skipped since Node.js became more strict :+1: 
- For anonymized proxy script Node.js 20 throw `socket hang up` instead of `ECONNREFUSED` for some special cases

### Overview

There was 2242 tests cases for Node.js v18. Currently there are 2191 (minus 51 test case) for Node.js v20. See details below:

1. "ignores non-standard server HTTP headers" - Skipped when `!useUpstreamProxy`:                                                                                                                                    
    - Main proxy configurations: 2 (http/https) × 2 (ssl) × 4 (auth) = 16 configs                                                                                                                                    
    - Custom response suites: 2 x 2 x 4 = 16 configs                                                                                                                                                                 
    - Direct tests (HTTP/HTTPS -> Target): 2 configs                                                                                                                                                                 
    - Total: 34 tests
2. "gracefully fails on invalid HTTP status code" - Skipped when `!useUpstreamProxy && !useSsl`:                                                                                                                     
    - Main proxy configurations: 2 x 1 x 4 = 8 configs                                                                                                                                                               
    - Custom response suites: 2 x 1 x 4 = 8 configs                                                                                                                                                                  
    - Direct tests: 1 config (HTTP -> Target only)                                                                                                                                                                   
    - Total: 17 tests
  
Grand total: 34 + 17 = 51 skipped tests

### Why these tests are skipped

Both tests verify the proxy's behavior when receiving malformed HTTP responses from upstream servers:
1. "ignores non-standard server HTTP headers" - tests handling of headers with spaces in names (e.g., Invalid Header With Space: value) which violate HTTP RFC
3. "gracefully fails on invalid HTTP status code" - tests handling of invalid status codes (like HTTP/1.1 55 OK)

In Node.js 20, the `--insecure-http-parser` flag no longer relaxes HTTP client parsing for these edge cases. The HTTP client throws parse errors before the proxy can even process them. This is a Node.js security hardening change, not a proxy bug.

Close #601

### CODE REVIEW

File: .github/workflows/check.yaml [hunk 1]
```diff
@@ -14,7 +14,7 @@ jobs:
 
     strategy:
       matrix:
-        node-version: [14, 16, 18]
+        node-version: [16, 18, 20]
```
Comment (jirimoravcik) [inline .github/workflows/check.yaml]: Why are we running checks on unsupported Node.js versions? Shouldn't this be `[20, 22, 24]`?

**Review by jirimoravcik — APPROVED** [review]

File: package.json [hunk 2]
```diff
@@ -44,7 +44,7 @@
         "lint:fix": "eslint . --fix"
     },
     "engines": {
-        "node": ">=14"
+        "node": ">=20"
```
Comment (danpoletaev) [inline package.json:47]: Shouldn't minimal node version be 16?
It should just warn user to use node 20, right?

Comment (danpoletaev) [reply package.json:47]: Let's also update `@types/node`

File: test/Dockerfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-ARG NODE_IMAGE=node:18.20.8-bookworm
+ARG NODE_IMAGE=node:20.19.6-bookworm
```
Comment (danpoletaev) [inline test/Dockerfile:1]: We need to also update `scripts/test-docker-all.sh`, since there're hardcoded node versions

**Review by danpoletaev — CHANGES_REQUESTED** [review]
Looks good!

Left few suggestions that needs to be reviewed before merging

File: .github/workflows/check.yaml (hunk 1 above)
Comment (bliuchak) [reply .github/workflows/check.yaml]: Good point. For some reason I kept this because all tests passed. But, you're probably right to drop old versions.

File: package.json (hunk 2 above)
Comment (bliuchak) [reply package.json:47]: The idea was to set minimal engine version to current oldest supported LTS version of Node. As of today it's Node 20.

Node 18 both active and security support period has ended.

File: test/Dockerfile (hunk 3 above)
Comment (bliuchak) [reply test/Dockerfile:1]: You're right. Gonna fix this.

File: .github/workflows/check.yaml (hunk 1 above)
Comment (bliuchak) [reply .github/workflows/check.yaml]: Fixed.

File: test/Dockerfile (hunk 3 above)
Comment (bliuchak) [reply test/Dockerfile:1]: Fixed

File: package.json (hunk 2 above)
Comment (bliuchak) [reply package.json:47]: > Let's also update @types/node

Good point. Upgraded.

**Review by danpoletaev — APPROVED** [review]
Nice ❤️

**Review by lewis-wow — APPROVED** [review]

**Comment by jancurn** [comment]
Hey guys, any blockers here? Let's ship this
